### ISSUE TITLE
Attempt to update object did not result in a single row modification

### ISSUE THREAD

**Issue by antbrown** [archivesspace/archivesspace#3105]
Hi there, our app is no longer starting and is stuck in an endless loop trying to grant permissions to a group. I wonder if there is anyway we can resolve this issue?

Error in logs (some info redacted, but you get the idea):

```
INFO: INFO: resetting rack response due exception: org.jruby.rack.RackInitializationException: 
Attempt to update object did not result in a single row modification (SQL: 
UPDATE `group` 
SET `lock_version` = 135, `json_schema_version` = 1, `repo_id` = 4, `group_code` = 'repository-managers', 
`group_code_norm` = 'repository-managers', `description` = 'Managers of the Test Repository', 
`created_by` = 'admin', `last_modified_by` = 'admin', `create_time` = '2019-03-14 21:23:34', 
`system_mtime` = '2024-01-26 16:27:52', `user_mtime` = '2019-03-14 21:23:34' 
WHERE ((`id` = 17) AND (`lock_version` = 134)) LIMIT 1
)
```

In mysql shell, if we try to find this row, it doesn't exist:

```
MariaDB [archivesspace]> select * from `group` where ((`id` = 17) AND (`lock_version` = 134));
Empty set (0.004 sec)
```

However, the row with id `17` exists, but its lock version has been updated:

```
MariaDB [archivesspace]> select * from `group` where id = 17;
+----+--------------+---------------------+---------+---------------------+---------------------+-----------------------------------------------------+------------+------------------+---------------------+---------------------+---------------------+
| id | lock_version | json_schema_version | repo_id | group_code          | group_code_norm     | description                                         | created_by | last_modified_by | create_time         | system_mtime        | user_mtime          |
+----+--------------+---------------------+---------+---------------------+---------------------+-----------------------------------------------------+------------+------------------+---------------------+---------------------+---------------------+
| 17 |          135 |                   1 |       4 | repository-managers | repository-managers | Managers of the Test Repository 15032019 repository | admin      | admin            | 2019-03-14 21:23:34 | 2024-01-26 16:27:52 | 2019-03-14 21:23:34 |
+----+--------------+---------------------+---------+---------------------+---------------------+-----------------------------------------------------+------------+------------------+---------------------+---------------------+---------------------+
1 row in set (0.001 sec)
```

I tried setting the lock_version back to 134, but it was immediately updated to 135 again and the error in the logs repeats ad nauseam.

It seems to be something to do with granting permissions to this group, as the stack trace refers to https://github.com/archivesspace/archivesspace/blob/v3.3.1/backend/app/lib/bootstrap_access_control.rb#L306

### Full stacktrace

```
INFO: INFO: resetting rack response due exception: org.jruby.rack.RackInitializationException: Attempt to update object did not result in a single row modification (SQL: UPDATE `group` SET `lock_version` = 135, `json_schema_version` = 1, `repo_id` = 4, `group_code` = 'repository-managers', `group_code_norm` = 'repository-managers', `description` = 'Managers of the Test Repository', `created_by` = 'admin', `last_modified_by` = 'admin', `create_time` = '2019-03-14 21:23:34', `system_mtime` = '2024-01-26 16:27:52', `user_mtime` = '2019-03-14 21:23:34' WHERE ((`id` = 17) AND (`lock_version` = 134)) LIMIT 1)
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1806:in `_update'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1799:in `_update_columns'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/plugins/optimistic_locking.rb:79:in `_update_columns'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1753:in `block in _save'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1013:in `around_update'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1735:in `block in _save'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1013:in `around_save'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/model/ASModel_sequel.rb:128:in `around_save'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1717:in `_save'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1446:in `block in save'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1864:in `block in checked_transaction'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/database/transactions.rb:175:in `block in transaction'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/connection_pool/threaded.rb:87:in `hold'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/database/connecting.rb:270:in `synchronize'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/database/transactions.rb:145:in `transaction'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1864:in `checked_transaction'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1446:in `block in save'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1852:in `checked_save_failure'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/base.rb:1446:in `save'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/lib/bootstrap_access_control.rb:314:in `block in set_up_base_permissions'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/dataset/actions.rb:149:in `block in each'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/adapters/jdbc.rb:795:in `process_result_set'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/adapters/jdbc.rb:726:in `block in fetch_rows'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/adapters/jdbc.rb:244:in `block in execute'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/adapters/jdbc.rb:675:in `statement'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/adapters/jdbc.rb:239:in `block in execute'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/connection_pool/threaded.rb:87:in `hold'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/database/connecting.rb:270:in `synchronize'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/adapters/jdbc.rb:238:in `execute'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/dataset/actions.rb:1082:in `execute'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/adapters/jdbc.rb:726:in `fetch_rows'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/dataset/actions.rb:149:in `each'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/model/plugins.rb:33:in `each'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/lib/bootstrap_access_control.rb:311:in `set_up_base_permissions'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/lib/bootstrap_access_control.rb:349:in `<class:ArchivesSpaceService>'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/lib/bootstrap_access_control.rb:1:in `<main>'
        from org/jruby/RubyKernel.java:974:in `require'
        from org/jruby/RubyKernel.java:1002:in `require_relative'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/main.rb:225:in `block in ArchivesSpaceService'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/model/db.rb:210:in `block in open'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/model/db.rb:101:in `block in transaction'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/database/transactions.rb:204:in `_transaction'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/database/transactions.rb:179:in `block in transaction'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/connection_pool/threaded.rb:91:in `hold'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/database/connecting.rb:270:in `synchronize'
        from /opt/archivesspace-3.3.1/gems/gems/sequel-5.9.0/lib/sequel/database/transactions.rb:145:in `transaction'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/model/db.rb:100:in `transaction'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/model/db.rb:207:in `block in open'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/model/db.rb:204:in `open'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/model/db.rb:542:in `block in open'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/main.rb:224:in `block in ArchivesSpaceService'
        from /opt/archivesspace-3.3.1/gems/gems/sinatra-2.0.5/lib/sinatra/base.rb:1426:in `configure'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/main.rb:101:in `<class:ArchivesSpaceService>'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/app/main.rb:48:in `<main>'
        from org/jruby/RubyKernel.java:974:in `require'
        from uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:83:in `require'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/config.ru:4:in `block in <main>'
        from org/jruby/RubyBasicObject.java:2622:in `instance_eval'
        from uri:classloader:/vendor/rack-2.2.3/rack/builder.rb:125:in `initialize'
        from /opt/archivesspace-3.3.1/data/tmp/jetty-0_0_0_0-8089-backend_war-_-any-3135108984332208751/webapp/WEB-INF/config.ru:1:in `<main>'
        from launcher/launcher.rb:90:in `start_server'
        from launcher/launcher.rb:144:in `main'
        from launcher/launcher.rb:242:in `<main>'
```

## Your Environment

* Version used: 3.3.1
* Environment name and version (e.g. Chrome 39, node.js 5.4): 
* Operating System and version (desktop or mobile): Ubuntu 22.04
* Link to your project:

**Comment by antbrown** [comment]
This issue has resolved itself and I don't know why... yesterday was a public holiday here in NZ and so no one was working on the site and the last log entry of the day was the same error as detailed above. This morning when the server woke up it just started normally with no mention of the issue.

**Comment by antbrown** [comment]
Hi, just on the off chance someone has encountered this issue I would be keen on hearing any ideas of how to resolve it... last time it fixed itself. Unfortunately it has occurred again just recently.

**Comment by brianzelip** [comment]
Hi @antbrown, sorry to hear about the issue you're experiencing. I'll share this with some folks to get their input and follow up here.

**Comment by jdshaw** [comment]
This issue thread in the Sequel gem repo indicates that concurrency may be a factor. https://github.com/jeremyevans/sequel/issues/1432

**Comment by antbrown** [comment]
Hi, thank you for the replies. That issue on sequel sounds very relevant. I'll let you know if I come up with any bright ideas that fix it for us :-)

**Comment by antbrown** [comment]
I found out this morning that whenever I was issuing the restart command it would silently fail to do anything. I had to ask a sysadmin to kill the process running ArchivesSpace and start fresh. This time it came up without error.

I'll close this issue (again), my gut feeling is that there is a problem with the script responsible for monitoring and bringing up ArchivesSpace if it is not running, perhaps it is trying to start multiple instances, I will keep an eye on it next time the server restarts.
